tests/unit/test_application.py

```diff
@@ -143,3 +145,69 @@ def test_build_planner_success_architecture_all(base, build_base, expected_base)
         assert build_info.platform == f"{expected_base.name}@{expected_base.version}"
 
     assert "all" not in [a.build_on for a in architectures]
+
+
+@pytest.fixture()
+def extension_source(default_project):
+    source = default_project.marshal()
+    source["confinement"] = "strict"
+    source["apps"] = {
+        "app1": {
+            "command": "app1",
+            "extensions": ["fake-extension"],
+        }
+    }
+    return source
+
+
+@pytest.mark.usefixtures("fake_extension")
+def test_application_expand_extensions(emitter, monkeypatch, extension_source, new_dir):
+    monkeypatch.setenv("CRAFT_DEBUG", "1")
+
+    (new_dir / "snap").mkdir()
+    (new_dir / "snap/snapcraft.yaml").write_text(json.dumps(extension_source))
+
+    monkeypatch.setattr("sys.argv", ["snapcraft", "expand-extensions"])
+    application.main()
+    emitter.assert_message(
+        dedent(
+            """\
+            name: default
+            version: '1.0'
+            summary: default project
+            description: default project
+            base: core24
+            build-base: devel
+            license: MIT
+            parts:
+                fake-extension/fake-part:
+                    plugin: nil
+            confinement: strict
+            grade: devel
+            apps:
+                app1:
+                    command: app1
+                    plugs:
+                    - fake-plug
+        """
+        )
+    )
+
+
+@pytest.mark.usefixtures("fake_extension")
+def test_application_build_with_extensions(monkeypatch, extension_source, new_dir):
+    """Test that extensions are correctly applied in regular builds."""
+    monkeypatch.setenv("CRAFT_DEBUG", "1")
```

Comment:
Was this for debugging the test during development?

Reply:
yeap. I'm going to leave it because it turns out it's _a bit useful_ to have the traceback when the test fails heh
we should enable that globally on the test runs probably